[NFC] Add a "StringLiteralInitializerInfo" utility class.

zoecarver opened the pull request:
This is the `StringLiteralInfo` utility class from #27551. I broke it out of that PR for ease of reviewing. I'm happy to move this into another place if that's better, too.

gottesmm commented on include/swift/SILOptimizer/Utils/ConstExpr.h:
```diff
@@ -40,6 +42,81 @@ class SymbolicValueAllocator;
+    if (makeStr->getNumArguments() < 1)
+      return {};
+
+    if (auto *SL = dyn_cast<StringLiteralInst>(makeStr->getOperand(1)))
```
How do you know this from a string init function?

gottesmm replied:
Also, can you change `SL` to be `sli`?

gottesmm commented on include/swift/SILOptimizer/Utils/ConstExpr.h:
```diff
@@ -40,6 +42,81 @@ class SymbolicValueAllocator;
+    if (makeStr->getNumArguments() < 3)
+      return false;
+
+    if (auto *isAscii = dyn_cast<IntegerLiteralInst>(makeStr->getOperand(3)))
```
Same question here.

gottesmm reviewed (commented):
Few questions.

zoecarver commented:
@gottesmm, thanks for the review! I think I fixed all your comments.

gottesmm commented on include/swift/SILOptimizer/Utils/ConstExpr.h:
```diff
@@ -40,6 +42,89 @@ class SymbolicValueAllocator;
 class ConstExprFunctionState;
 class UnknownReason;
 
+class StringLiteralInfo {
```
Can you add a nice block comment to this describing the class?

gottesmm commented on include/swift/SILOptimizer/Utils/ConstExpr.h:
```diff
@@ -40,6 +42,89 @@ class SymbolicValueAllocator;
+    return {};
+  }
+
+  static Optional<StringLiteralInfo> create(SILInstruction *inst) {
```
Can you add a comment here and all unadorned functions/fields here?

gottesmm commented on include/swift/SILOptimizer/Utils/ConstExpr.h:
```diff
@@ -40,6 +42,89 @@ class SymbolicValueAllocator;
+  /// If the given instruction is a call to the compiler-intrinsic initializer
+  /// of String that accepts string literals, return the called function.
+  /// Otherwise, return nullptr.
+  static SILFunction *getStringMakeUTF8Init(SILInstruction *inst) {
```
I would be even clearer here about what that function is, what you expect its signature to be, etc.

gottesmm replied:
@zoecarver 

My suggestion: constructFromStringMakeUTF8Init(...)?.

zoecarver replied:
@gottesmm what do you mean? Are you suggesting that as the function name? I'll add a bit more clarity about the function signature.

zoecarver replied:
Maybe `getStringMakeUTF8InitFunction` would be better because this returns a `SILFunction*`, it doesn't construct a `StringLiteralInfo` object.

gottesmm commented on include/swift/SILOptimizer/Utils/ConstExpr.h:
```diff
@@ -40,6 +42,89 @@ class SymbolicValueAllocator;
+    if (!apply)
+      return nullptr;
+
+    SILFunction *callee = apply->getCalleeFunction();
```
Couldn't getStringMakeUTF8Init be implemented in terms of this?

zoecarver replied:
Yes, but then we would have to call `getCalleeFunction` twice.

gottesmm commented on include/swift/SILOptimizer/Utils/ConstExpr.h:
```diff
@@ -40,6 +42,89 @@ class SymbolicValueAllocator;
+  }
+
+  /// Gets an optional StringRef from a string init function.
+  static Optional<StringRef> getUTF8String(ApplyInst *makeStr) {
```
Can you elaborate here in the comment about what you are pattern matching. I think something like this would probably be clearer:

"If \p makeStr is an ApplyInst that calls a function with the semantics attribute 'string.makeUTF8', if its first argument is a StringLiteralInst, return the contents of the StringLiteralInst as a StringRef"

gottesmm commented on include/swift/SILOptimizer/Utils/ConstExpr.h:
```diff
@@ -40,6 +42,89 @@ class SymbolicValueAllocator;
+  }
+
+  /// Gets is string ascii from string init function.
+  static bool getIsAscii(ApplyInst *makeStr) {
```
I dislike how this code is assuming that string.makeUTF8 has a certain form. I wonder if we could put in an assert of some sort that checks that any function with string.makeUTF8 that this is called upon has the proper form? Not sure if it is worth it. Definitely though the expected functional form should be documented here.

Otherwise, can you improve this comment along the lines of the one above the StringLiteralInst.

zoecarver replied:
I suppose I could assert that the label is correct (`isASCII`) but, I think it's probably OK with a comment on it that describes how the function should look.

zoecarver commented:
@gottesmm friendly ping. Any other comments?

gottesmm commented:
Please rebase before I look.

zoecarver commented:
@gottesmm rebased, thanks for taking a look.

gottesmm commented on include/swift/AST/SemanticAttrs.h:
```diff
@@ -23,7 +23,8 @@
 namespace semantics {
-#define SEMANTICS_ATTR(NAME, C_STR) constexpr static const StringLiteral NAME = C_STR;
+#define SEMANTICS_ATTR(NAME, C_STR) \
+  constexpr static const llvm::StringLiteral NAME = C_STR;
```
StringLiteral is injected into the swift namespace by https://github.com/apple/swift/blob/master/include/swift/Basic/LLVM.h. Maybe just include that and get rid of llvm::?

gottesmm commented on include/swift/SILOptimizer/Utils/ConstExpr.h:
```diff
@@ -40,6 +43,106 @@ class SymbolicValueAllocator;
+  StringRef value;
+
+  /// Create an instance of StringLiteralInfo by casting v to an ApplyInst.
+  /// \param v SILValue to be cast to ApplyInst and used in creation of the string literal.
```
One ask. Can you put a newline in between "Create an instance" and "\param v". Can you do something similar in the rest of the changes?

gottesmm reviewed (commented):
some more thoughts. I am still nervous about the names of the static constructors.

zoecarver commented:
@gottesmm I'd really like to commit this patch so other PRs like #28260 can use it. Any other comments?

futurejones commented:
Can one of the admins verify this patch?

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

zoecarver committed:
[NFC] Add a "StringLiteralInitializerInfo" utility class.

Add "StringLiteralInitializerInfo," a utility class to help analyze
string literal initializer functions.
